signalling/src/server/mod.rs

```diff
@@ -0,0 +1,227 @@
+            }
+
+            ws_sink.send(WsMessage::Close(None)).await?;
+            ws_sink.close().await?;
+
+            Ok::<(), Error>(())
+        });
+
+        {}
+
+        let state_clone = state.clone();
+        let this_id_clone = this_id.clone();
+        let receive_task_handle = task::spawn(async move {
+            while let Some(msg) = async_std::stream::StreamExt::next(&mut ws_stream).await {
+                let mut state = state_clone.lock().unwrap();
+
+                match msg {
+                    Ok(WsMessage::Text(msg)) => {
+                        let mut message_sender = state.message_sender.take().unwrap();
+
+                        if let Err(err) = state.message_handler.handle_message(
+                            &mut message_sender,
+                            msg,
+                            &this_id_clone,
+                        ) {
+                            warn!(this = %this_id_clone, "Error handling message: {:?}", err);
+                            message_sender.send_message(
+                                serde_json::to_string(&p::OutgoingMessage::Error {
+                                    details: err.to_string(),
+                                })
+                                .unwrap(),
+                                &this_id_clone,
+                            );
+                        }
+
+                        state.message_sender = Some(message_sender);
+                    }
+                    Ok(WsMessage::Close(reason)) => {
+                        info!(this_id = %this_id_clone, "connection closed: {:?}", reason);
+                        break;
+                    }
+                    Ok(_) => warn!(this_id = %this_id_clone, "Unsupported message type"),
+                    Err(err) => {
+                        warn!(this_id = %this_id_clone, "recv error: {}", err);
+                        break;
+                    }
+                }
+            }
+
+            Self::remove_peer(state_clone, &this_id_clone);
+        });
+
+        let mut state = state.lock().unwrap();
+
+        state.message_sender.as_mut().unwrap().peers.insert(
+            this_id,
+            Peer {
+                receive_task_handle,
+                send_task_handle,
+                sender: websocket_sender,
+            },
+        );
+    }
+
+    /// Run the server
+    pub async fn run(&self, host: &str, port: u16) -> Result<(), SignallingServerError> {
+        let addr = format!("{}:{}", host, port);
+
+        // Create the event loop and TCP listener we'll accept connections on.
+        let try_socket = TcpListener::bind(&addr).await;
+        let listener = try_socket.expect("Failed to bind");
+        info!("Listening on: {}", addr);
+
+        let state_clone = self.state.clone();
+        while let Ok((stream, _)) = listener.accept().await {
+            task::spawn(Self::accept_connection(state_clone.clone(), stream));
```

Comment:
It might make sense to also make `Self::accept_connection` public. Then the application can do its connection setup itself and in the end pass the connection here for being taken care of further.